src/System.Net.NameResolution/src/System/Net/DNS.cs

```diff
@@ -0,0 +1,383 @@
+            IAsyncResult asyncResult = HostResolutionBeginHelper(hostNameOrAddress, false, requestCallback, stateObject);
+
+            if (Logging.On) Logging.Exit(Logging.Sockets, "DNS", "BeginGetHostEntry", asyncResult);
+            return asyncResult;
+        } // BeginResolve
+
+        public static IAsyncResult BeginGetHostEntry(IPAddress address, AsyncCallback requestCallback, object stateObject)
+        {
+            if (Logging.On) Logging.Enter(Logging.Sockets, "DNS", "BeginGetHostEntry", address);
+            IAsyncResult asyncResult = HostResolutionBeginHelper(address, true, true, requestCallback, stateObject);
+
+            if (Logging.On) Logging.Exit(Logging.Sockets, "DNS", "BeginGetHostEntry", asyncResult);
+            return asyncResult;
+        } // BeginResolve
+
+        public static IPHostEntry EndGetHostEntry(IAsyncResult asyncResult)
+        {
+            if (Logging.On) Logging.Enter(Logging.Sockets, "DNS", "EndGetHostEntry", asyncResult);
+            IPHostEntry ipHostEntry = HostResolutionEndHelper(asyncResult);
+
+            if (Logging.On) Logging.Exit(Logging.Sockets, "DNS", "EndGetHostEntry", ipHostEntry);
+            return ipHostEntry;
+        } // EndResolve()
+
+        public static IAsyncResult BeginGetHostAddresses(string hostNameOrAddress, AsyncCallback requestCallback, object state)
+        {
+            if (Logging.On) Logging.Enter(Logging.Sockets, "DNS", "BeginGetHostAddresses", hostNameOrAddress);
+            IAsyncResult asyncResult = HostResolutionBeginHelper(hostNameOrAddress, true, requestCallback, state);
+
+            if (Logging.On) Logging.Exit(Logging.Sockets, "DNS", "BeginGetHostAddresses", asyncResult);
+            return asyncResult;
+        } // BeginResolve
+
+        public static IPAddress[] EndGetHostAddresses(IAsyncResult asyncResult)
+        {
+            if (Logging.On) Logging.Enter(Logging.Sockets, "DNS", "EndGetHostAddresses", asyncResult);
+            IPHostEntry ipHostEntry = HostResolutionEndHelper(asyncResult);
+
+            if (Logging.On) Logging.Exit(Logging.Sockets, "DNS", "EndGetHostAddresses", ipHostEntry);
+            return ipHostEntry.AddressList;
+        } // EndResolveToAddresses
+
+        //************* Task-based async public methods *************************
+        public static Task<IPAddress[]> GetHostAddressesAsync(string hostNameOrAddress)
+        {
+            return Task<IPAddress[]>.Factory.FromAsync(BeginGetHostAddresses, EndGetHostAddresses, hostNameOrAddress, null);
+        }
+
+        public static Task<IPHostEntry> GetHostEntryAsync(IPAddress address)
+        {
+            return Task<IPHostEntry>.Factory.FromAsync(BeginGetHostEntry, EndGetHostEntry, address, null);
+        }
+
+        public static Task<IPHostEntry> GetHostEntryAsync(string hostNameOrAddress)
+        {
+            return Task<IPHostEntry>.Factory.FromAsync(BeginGetHostEntry, EndGetHostEntry, hostNameOrAddress, null);
+        }
```

Comment:
Actually, I just realized that these are all static methods... that means that it's actually much easier to avoid the allocations.  A short term change to avoid the delegate allocations would look like this, basically just using lambdas rather than method group delegate inference, because the compiler is able to cache the former but currently can't cache the latter.
```C#
public static Task<IPAddress[]> GetHostAddressesAsync(string hostNameOrAddress)
{
    return Task<IPAddress[]>.Factory.FromAsync(
        (arg, requestCallback, stateObject) => BeginGetHostAddresses(arg, requestCallback, stateObject),
        asyncResult => EndGetHostAddresses(asyncResult),
        hostNameOrAddress,
        null);
}

public static Task<IPHostEntry> GetHostEntryAsync(IPAddress address)
{
    return Task<IPHostEntry>.Factory.FromAsync(
        (arg, requestCallback, stateObject) => BeginGetHostEntry(arg, requestCallback, stateObject), 
        asyncResult => EndGetHostEntry(asyncResult), 
        address, 
        null);
}

public static Task<IPHostEntry> GetHostEntryAsync(string hostNameOrAddress)
{
    return Task<IPHostEntry>.Factory.FromAsync(
        (arg, requestCallback, stateObject) => BeginGetHostEntry(arg, requestCallback, stateObject),
        asyncResult => EndGetHostEntry(asyncResult),
        hostNameOrAddress,
        null);
}
```
That's probably a worthwhile change to make now for the short-term.  The longer term improvements, both here and in other Task implementations, is to not have the Task-based APIs be wrappers for the APM ones. For example, GetHostAddressesAsync creates a Task to wrap the Begin/End methods, which in turn schedule a Task... it'd be better to skip the middle-men.  I'll open a work item for that.